// @qlik/sdk/dist/esm/qlik.js
import { ApiKeys } from "./apis/Api_Keys.js";
import { Apps } from "./apis/Apps.js";
import { Audits } from "./apis/Audits.js";
import { Automations } from "./apis/Automations.js";
import { Collections } from "./apis/Collections.js";
import { CspOrigins } from "./apis/Csp_Origins.js";
import { DataFiles } from "./apis/Data_Files.js";
import { Extensions } from "./apis/Extensions.js";
import { Groups } from "./apis/Groups.js";
import { IdentityProviders } from "./apis/Identity_Providers.js";
import { Items } from "./apis/Items.js";
import { Licenses } from "./apis/Licenses.js";
import { Notes } from "./apis/Notes.js";
import { OauthTokens } from "./apis/Oauth_Tokens.js";
import { Questions } from "./apis/Questions.js";
import { Quotas } from "./apis/Quotas.js";
import { ReloadTasks } from "./apis/Reload_Tasks.js";
import { Reloads } from "./apis/Reloads.js";
import { Roles } from "./apis/Roles.js";
import { Spaces } from "./apis/Spaces.js";
import { Themes } from "./apis/Themes.js";
import { Transports } from "./apis/Transports.js";
import { Users } from "./apis/Users.js";
import { WebIntegrations } from "./apis/Web_Integrations.js";
import { Webhooks } from "./apis/Webhooks.js";
import Auth from "./auth/auth.js";
import { ApiKeys as ApiKeys2 } from "./apis/Api_Keys.js";
import { Apps as Apps2 } from "./apis/Apps.js";
import { Audits as Audits2 } from "./apis/Audits.js";
import { Automations as Automations2 } from "./apis/Automations.js";
import { Collections as Collections2 } from "./apis/Collections.js";
import { CspOrigins as CspOrigins2 } from "./apis/Csp_Origins.js";
import { DataFiles as DataFiles2 } from "./apis/Data_Files.js";
import { Extensions as Extensions2 } from "./apis/Extensions.js";
import { Groups as Groups2 } from "./apis/Groups.js";
import { IdentityProviders as IdentityProviders2 } from "./apis/Identity_Providers.js";
import { Items as Items2 } from "./apis/Items.js";
import { Licenses as Licenses2 } from "./apis/Licenses.js";
import { Notes as Notes2 } from "./apis/Notes.js";
import { OauthTokens as OauthTokens2 } from "./apis/Oauth_Tokens.js";
export * from "./apis/Qix.js";
import { Questions as Questions2 } from "./apis/Questions.js";
import { Quotas as Quotas2 } from "./apis/Quotas.js";
import { ReloadTasks as ReloadTasks2 } from "./apis/Reload_Tasks.js";
import { Reloads as Reloads2 } from "./apis/Reloads.js";
import { Roles as Roles2 } from "./apis/Roles.js";
import { Spaces as Spaces2 } from "./apis/Spaces.js";
import { Themes as Themes2 } from "./apis/Themes.js";
import { Transports as Transports2 } from "./apis/Transports.js";
import { Users as Users2 } from "./apis/Users.js";
import { WebIntegrations as WebIntegrations2 } from "./apis/Web_Integrations.js";
import { Webhooks as Webhooks2 } from "./apis/Webhooks.js";
import { AuthType } from "./types/types.js";
import {
  ConnectionErrorMsg,
  AuthenticationErrorMsg,
  CustomErrorMsg
} from "./errors.js";
class Qlik {
  constructor(config) {
    this.auth = new Auth(config);
    this.rest = this.auth.rest;
    this.rpc = this.auth.rpc;
    this.csporigins = new CspOrigins(this.auth);
    this.datafiles = new DataFiles(this.auth);
    this.webintegrations = new WebIntegrations(this.auth);
    this.apiKeys = new ApiKeys(this.auth);
    this.apps = new Apps(this.auth);
    this.audits = new Audits(this.auth);
    this.automations = new Automations(this.auth);
    this.collections = new Collections(this.auth);
    this.cspOrigins = new CspOrigins(this.auth);
    this.dataFiles = new DataFiles(this.auth);
    this.extensions = new Extensions(this.auth);
    this.groups = new Groups(this.auth);
    this.identityProviders = new IdentityProviders(this.auth);
    this.items = new Items(this.auth);
    this.licenses = new Licenses(this.auth);
    this.notes = new Notes(this.auth);
    this.oauthTokens = new OauthTokens(this.auth);
    this.questions = new Questions(this.auth);
    this.quotas = new Quotas(this.auth);
    this.reloadTasks = new ReloadTasks(this.auth);
    this.reloads = new Reloads(this.auth);
    this.roles = new Roles(this.auth);
    this.spaces = new Spaces(this.auth);
    this.themes = new Themes(this.auth);
    this.transports = new Transports(this.auth);
    this.users = new Users(this.auth);
    this.webIntegrations = new WebIntegrations(this.auth);
    this.webhooks = new Webhooks(this.auth);
  }
}
export {
  ApiKeys2 as ApiKeys,
  Apps2 as Apps,
  Audits2 as Audits,
  Auth,
  AuthType,
  AuthenticationErrorMsg,
  Automations2 as Automations,
  Collections2 as Collections,
  ConnectionErrorMsg,
  CspOrigins2 as CspOrigins,
  CustomErrorMsg,
  DataFiles2 as DataFiles,
  Extensions2 as Extensions,
  Groups2 as Groups,
  IdentityProviders2 as IdentityProviders,
  Items2 as Items,
  Licenses2 as Licenses,
  Notes2 as Notes,
  OauthTokens2 as OauthTokens,
  Questions2 as Questions,
  Quotas2 as Quotas,
  ReloadTasks2 as ReloadTasks,
  Reloads2 as Reloads,
  Roles2 as Roles,
  Spaces2 as Spaces,
  Themes2 as Themes,
  Transports2 as Transports,
  Users2 as Users,
  WebIntegrations2 as WebIntegrations,
  Webhooks2 as Webhooks,
  Qlik as default
};
