// @qlik/sdk/dist/cjs/apis/Groups.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Groups_exports = {};
__export(Groups_exports, {
  Group: () => Group,
  Groups: () => Groups
});
module.exports = __toCommonJS(Groups_exports);
var import_auth = __toESM(require("../auth/auth.js"));
var import_listable = __toESM(require("../listable.js"));
class Group {
  constructor({
    assignedRoles,
    createdAt,
    id,
    idpId,
    lastUpdatedAt,
    links,
    name,
    status,
    tenantId,
    ...rest
  }, auth) {
    this.auth = auth;
    if (assignedRoles !== void 0 && assignedRoles !== null) {
      this.assignedRoles = assignedRoles;
    }
    this.createdAt = createdAt;
    this.id = id;
    this.idpId = idpId;
    this.lastUpdatedAt = lastUpdatedAt;
    this.links = links;
    this.name = name;
    this.status = status;
    this.tenantId = tenantId;
    if (rest) {
      Object.keys(rest).forEach((prop) => {
        this[prop] = rest[prop];
      });
    }
  }
  /**
    Delete group by id
  
    */
  async delete() {
    await this.auth.rest(
      "/groups/{groupId}".replace("{groupId}", this.id ?? ""),
      {
        method: "DELETE"
      }
    );
  }
  /**
    Update group by id
  
      @param data - An array of JSON Patches for a group.
    */
  async patch(data) {
    await this.auth.rest(
      "/groups/{groupId}".replace("{groupId}", this.id ?? ""),
      {
        method: "PATCH",
        body: JSON.stringify(data)
      }
    );
  }
}
class Groups {
  constructor(config) {
    this.config = config instanceof import_auth.default ? config.config : config;
    this.auth = config instanceof import_auth.default ? config : new import_auth.default(config);
  }
  /**    Retrieves a list of groups matching the filter using advanced query string.
    Filter groups
  
      @param limit - The number of user entries to retrieve.
      @param next - Get users with IDs that are higher than the target user ID. Cannot be used in conjunction with prev.
      @param prev - Get users with IDs that are lower than the target user ID. Cannot be used in conjunction with next.
      @param sort - The field to sort by, with +/- prefix indicating sort order
      @param data - An advanced query filter to be used for complex user querying in the tenant.
    */
  async filter(data, queryParams = { limit: 20, sort: "+name" }) {
    const response = await this.auth.rest(
      `/groups/actions/filter?${new URLSearchParams(Object.entries(queryParams).filter(([, v]) => v !== void 0).map(([k, v]) => [k, v.toString()])).toString()}`,
      {
        method: "POST",
        body: JSON.stringify(data)
      }
    );
    const responseJson = await response.json();
    return new import_listable.default(responseJson, this.auth, Group, queryParams);
  }
  /**    Returns the active tenant's group settings.
    Get group settings
  
    */
  async getSettings() {
    const response = await this.auth.rest(
      "/groups/settings",
      {
        method: "GET"
      }
    );
    return response.json();
  }
  /**
    Update group settings
  
      @param data - An array of JSON Patches for the groups settings.
    */
  async patchSettings(data) {
    await this.auth.rest(
      "/groups/settings",
      {
        method: "PATCH",
        body: JSON.stringify(data)
      }
    );
  }
  /**    Returns the requested group.
    Get group by id
  
      @param groupId - The group's unique identifier
    */
  async get(groupId) {
    const response = await this.auth.rest(
      "/groups/{groupId}".replace("{groupId}", groupId),
      {
        method: "GET"
      }
    );
    const responseJson = await response.json();
    return new Group(responseJson, this.auth);
  }
  /**    Returns a list of groups with cursor-based pagination.
    List groups.
  
      @param filter - The advanced filtering to use for the query. Refer to RFC 7644 https://datatracker.ietf.org/doc/rfc7644/ for the syntax. Cannot be combined with any of the fields marked as deprecated. All conditional statements within this query parameter are case insensitive.
      @param limit - The number of groups to retrieve.
      @param next - The next page cursor.
      @param prev - The previous page cursor.
      @param sort - Optional resource field name to sort on, eg. name. Can be prefixed with +/- to determine order, defaults to (+) ascending.
      @param systemGroups - Return system groups (e.g. Everyone) instead of regular groups. Cannot be combined with any other query parameters.
      @param totalResults - Whether to return a total match count in the result. Defaults to false.
    */
  async getGroups(queryParams = { limit: 20 }) {
    const response = await this.auth.rest(
      `/groups?${new URLSearchParams(Object.entries(queryParams).filter(([, v]) => v !== void 0).map(([k, v]) => [k, v.toString()])).toString()}`,
      {
        method: "GET"
      }
    );
    const responseJson = await response.json();
    return new import_listable.default(responseJson, this.auth, Group, queryParams);
  }
  /**    Creates a group. The maximum number of groups a tenant can have is 10,000.
    Create group.
  
      @param data -
    */
  async create(data) {
    const response = await this.auth.rest(
      "/groups",
      {
        method: "POST",
        body: JSON.stringify(data)
      }
    );
    const responseJson = await response.json();
    return new Group(responseJson, this.auth);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Group,
  Groups
});
