// @qlik/sdk/dist/mjs/apis/Licenses.js
import Auth from "../auth/auth.js";
import ListableResource from "../listable.js";
class Licenses {
  constructor(config) {
    this.config = config instanceof Auth ? config.config : config;
    this.auth = config instanceof Auth ? config : new Auth(config);
  }
  /**
    Assigns license access to the given users
  
      @param data - List of subjects to allocate assignments for.
    */
  async add(data) {
    const response = await this.auth.rest(
      "/licenses/assignments/actions/add",
      {
        method: "POST",
        body: JSON.stringify(data)
      }
    );
    return response.json();
  }
  /**
    Removes license access for the given users
  
      @param data - List of assignments to delete.
    */
  async delete(data) {
    const response = await this.auth.rest(
      "/licenses/assignments/actions/delete",
      {
        method: "POST",
        body: JSON.stringify(data)
      }
    );
    return response.json();
  }
  /**
    Updates license access for the given users
  
      @param data - List of assignments to update.
    */
  async update(data) {
    const response = await this.auth.rest(
      "/licenses/assignments/actions/update",
      {
        method: "POST",
        body: JSON.stringify(data)
      }
    );
    return response.json();
  }
  /**
    Retrieves assignments for the current tenant
  
      @param filter - The filter for finding entries.
      @param limit - The preferred number of entries to return.
      @param page - The requested page.
      @param sort - The field to sort on; can be prefixed with +/- for ascending/descending sort order.
    */
  async getAssignments(queryParams = { limit: 20 }) {
    const response = await this.auth.rest(
      `/licenses/assignments?${new URLSearchParams(Object.entries(queryParams).filter(([, v]) => v !== void 0).map(([k, v]) => [k, v.toString()])).toString()}`,
      {
        method: "GET"
      }
    );
    const responseJson = await response.json();
    return new ListableResource(responseJson, this.auth, null, queryParams);
  }
  /**
    Retrieves license consumption for the current tenant
  
      @param filter - The filter for finding entries.
      @param limit - The preferred number of entries to return.
      @param page - The requested page.
      @param sort - The field to sort on; can be prefixed with +/- for ascending/descending sort order.
    */
  async getConsumptions(queryParams = { limit: 200 }) {
    const response = await this.auth.rest(
      `/licenses/consumption?${new URLSearchParams(Object.entries(queryParams).filter(([, v]) => v !== void 0).map(([k, v]) => [k, v.toString()])).toString()}`,
      {
        method: "GET"
      }
    );
    const responseJson = await response.json();
    return new ListableResource(responseJson, this.auth, null, queryParams);
  }
  /**
    Gets the general information of the license applied to the current tenant
  
    */
  async getOverview() {
    const response = await this.auth.rest(
      "/licenses/overview",
      {
        method: "GET"
      }
    );
    return response.json();
  }
  /**
    Get auto assign settings for tenant.
  
    */
  async getSettings() {
    const response = await this.auth.rest(
      "/licenses/settings",
      {
        method: "GET"
      }
    );
    return response.json();
  }
  /**
    Set auto assign settings for tenant
  
      @param data - Dynamic assignment settings for professional and analyzer users. If professional users and analyzer users are both set, professional users will be automatically assigned, if available. Otherwise, analyzer users will be assigned. If neither of those users are available, analyzer capacity will be assigned, if available.
    */
  async setSettings(data) {
    const response = await this.auth.rest(
      "/licenses/settings",
      {
        method: "PUT",
        body: JSON.stringify(data)
      }
    );
    return response.json();
  }
  /**
    Gets the license status information of the current tenant
  
    */
  async getStatus() {
    const response = await this.auth.rest(
      "/licenses/status",
      {
        method: "GET"
      }
    );
    return response.json();
  }
}
export {
  Licenses
};
