// @qlik/sdk/dist/mjs/apis/Transports.js
import Auth from "../auth/auth.js";
class Transports {
  constructor(config) {
    this.config = config instanceof Auth ? config.config : config;
    this.auth = config instanceof Auth ? config : new Auth(config);
  }
  /**
    Send a test mail with the supplied email info (subject, body, recipient). Email config from database is used for the connection.
  
      @param data -
    */
  async sendTestEmail(data) {
    const response = await this.auth.rest(
      "/transports/email-config/actions/send-test-email",
      {
        method: "POST",
        body: JSON.stringify(data)
      }
    );
    return response.json();
  }
  /**
    Returns the isValid value for the email configuration for the tenant. Will return false if no email configuration exists.
  
    */
  async validate() {
    const response = await this.auth.rest(
      "/transports/email-config/actions/validate",
      {
        method: "POST"
      }
    );
    return response.json();
  }
  /**
    Verifies connection to email server for tenant provided via JWT
  
    */
  async verifyConnection() {
    const response = await this.auth.rest(
      "/transports/email-config/actions/verify-connection",
      {
        method: "POST"
      }
    );
    return response.json();
  }
  /**
    Deletes email configuration for a given tenant id (retrieved from JWT).
  
    */
  async deleteEmailConfig() {
    await this.auth.rest(
      "/transports/email-config",
      {
        method: "DELETE"
      }
    );
  }
  /**
    Returns the email configuration for a given tenant id (retrieved from JWT).
  
    */
  async getEmailConfig() {
    const response = await this.auth.rest(
      "/transports/email-config",
      {
        method: "GET"
      }
    );
    return response.json();
  }
  /**
    Patch the email configuration for a given tenant id (retrieved from JWT).
  
      @param data -
    */
  async patchEmailConfig(data) {
    await this.auth.rest(
      "/transports/email-config",
      {
        method: "PATCH",
        body: JSON.stringify(data)
      }
    );
  }
}
export {
  Transports
};
