// @qlik/sdk/dist/mjs/auth/rpcClient.js

import { createRequire } from 'module';
const require = createRequire(import.meta.url);
import enigma from "enigma.js";
import {
  AuthType
} from "../types/types.js";
import RestClient from "./restClient.js";
import { generateWebsocketUrl, isBrowser } from "../utils/utils.js";
import { EngineCloseError, EngineMessageError, NotFoundError } from "../errors.js";
const WS = true ? require("ws") : WebSocket;
function use(intercept) {
  const interceptArr = Array.isArray(intercept) ? intercept : [intercept];
  this.handlers = this.handlers.concat(interceptArr);
}
class RpcClient {
  constructor(config, rest) {
    /** object storing the interceptors */
    this.interceptors = {
      request: {
        handlers: [],
        use
      },
      response: {
        handlers: [],
        use
      }
    };
    this.sessions = {};
    this.config = config;
    this.rest = rest ?? new RestClient(config).restClientInstance;
    const errorInterceptor = {
      onFulfilled: (session, req, res) => {
        if ("error" in res) {
          const { error } = res;
          throw new EngineMessageError(error.code, error.message, error.parameter);
        }
        return res;
      }
    };
    this.interceptors.response.use(errorInterceptor);
  }
  async rpc(appId, rpcConfig = { delta: true }) {
    const apiResponseInterceptor = {
      onFulfilled: (session, req, res) => {
        const qReturn = res?.result?.qReturn;
        if (qReturn && qReturn.qHandle && qReturn.qType) {
          res.result.qReturn.id = qReturn.qGenericId;
          res.result.qReturn.handle = qReturn.qHandle;
          res.result.qReturn.type = qReturn.qType;
          res.result.qReturn.genericType = qReturn.qGenericType;
        }
        if (qReturn && qReturn.qHandle === null && qReturn.qType === null) {
          const code = -2;
          throw new NotFoundError({ error: "Object not found", code });
        }
        return res;
      }
    };
    const { url, createSocket } = await this.buildUrlAndCreateSocket(appId);
    const createSession = () => {
      const session = enigma.create({
        schema: {
          structs: {
            Global: {},
            Doc: {},
            GenericMeasure: {},
            GenericVariable: {},
            GenericBookmark: {},
            GenericDimension: {},
            GenericObject: {},
            Field: {}
          }
        },
        url,
        createSocket,
        protocol: { delta: rpcConfig.delta }
      });
      this.interceptors.response.use(session.intercept.response[1]);
      this.interceptors.response.use(apiResponseInterceptor);
      session.intercept.response = this.interceptors.response.handlers;
      this.interceptors.request.use(session.intercept.request[0]);
      session.intercept.request = this.interceptors.request.handlers;
      session.on("closed", (data) => {
        if (this.sessions[session.config?.url]?.eventListeners) {
          Object.values(this.sessions[session.config?.url]?.eventListeners)?.forEach((handle) => {
            handle?.closed?.forEach(
              (listener) => {
                listener(data);
              }
            );
          });
        }
        if (data?.code && data.code >= 4e3) {
          session.onRpcError(new EngineCloseError(data.code, data.reason));
        }
        delete this.sessions[session.config?.url];
      });
      return session;
    };
    this.sessions[url.toString()] = this.sessions[url.toString()] || createSession();
    this.sessions[url.toString()].on("traffic:*", (direction, msg) => {
      if ("handle" in msg) {
        this.sessions[url.toString()]?.eventListeners?.[msg.handle]?.[`traffic:${direction}`]?.forEach(async (listener) => {
          await listener(msg);
        });
        this.sessions[url.toString()]?.eventListeners?.[msg.handle]?.["traffic:*"]?.forEach(async (listener) => {
          await listener(msg);
        });
      }
      if ("change" in msg) {
        msg.change.forEach((qHandle) => {
          this.sessions[url.toString()]?.eventListeners?.[qHandle]?.changed?.forEach(
            async (listener) => {
              await listener(msg);
            }
          );
        });
      }
      if ("close" in msg) {
        msg.close.forEach((qHandle) => {
          this.sessions[url.toString()]?.eventListeners?.[qHandle]?.closed?.forEach(
            async (listener) => {
              await listener(msg);
            }
          );
        });
      }
    });
    return this.sessions[url.toString()];
  }
  /**
  * @param appId - the application Id
  * @param config - required configuration
  */
  async buildUrlAndCreateSocket(appId) {
    const url = await generateWebsocketUrl(appId, this.rest, this.config);
    let createSocket;
    if (this.config.authType === AuthType.APIKey || this.config.authType === AuthType.OAuth2 && !isBrowser) {
      createSocket = (wsUrl) => new WS(wsUrl, {
        headers: { Authorization: `Bearer ${this.config.apiKey}` }
      });
    } else {
      createSocket = (wsUrl) => new WS(wsUrl);
    }
    return { url, createSocket };
  }
}
export {
  RpcClient as default
};
