// @qlik/sdk/dist/esm/apis/Reload_Tasks.js
import Auth from "../auth/auth.js";
import ListableResource from "../listable.js";
class Task {
  constructor({
    appId,
    autoReload,
    autoReloadPartial,
    endDateTime,
    partial,
    recurrence,
    startDateTime,
    timeZone,
    fortressId,
    id,
    lastExecutionTime,
    links,
    log,
    nextExecutionTime,
    spaceId,
    state,
    tenantId,
    userId,
    ...rest
  }, auth) {
    this.auth = auth;
    this.appId = appId;
    this.autoReload = autoReload;
    this.autoReloadPartial = autoReloadPartial;
    this.endDateTime = endDateTime;
    this.partial = partial;
    this.recurrence = recurrence;
    this.startDateTime = startDateTime;
    this.timeZone = timeZone;
    this.fortressId = fortressId;
    this.id = id;
    this.lastExecutionTime = lastExecutionTime;
    this.links = links;
    this.log = log;
    this.nextExecutionTime = nextExecutionTime;
    this.spaceId = spaceId;
    this.state = state;
    this.tenantId = tenantId;
    this.userId = userId;
    if (rest) {
      Object.keys(rest).forEach((prop) => {
        this[prop] = rest[prop];
      });
    }
  }
  /**
    Delete a task.
  
    */
  async delete() {
    await this.auth.rest(
      "/reload-tasks/{taskId}".replace("{taskId}", this.id ?? ""),
      {
        method: "DELETE"
      }
    );
  }
  /**
    Update an existing task.
  
      @param data - Request body specifying the task parameters.
    */
  async set(data) {
    const response = await this.auth.rest(
      "/reload-tasks/{taskId}".replace("{taskId}", this.id ?? ""),
      {
        method: "PUT",
        body: JSON.stringify(data)
      }
    );
    const responseJson = await response.json();
    Object.keys(responseJson).forEach((key) => {
      if (this[key]) {
        this[key] = responseJson[key];
      }
    });
    return this;
  }
}
class ReloadTasks {
  constructor(config) {
    this.config = config instanceof Auth ? config.config : config;
    this.auth = config instanceof Auth ? config : new Auth(config);
  }
  /**
    Find and return a task.
  
      @param taskId - The unique identifier of the task.
    */
  async get(taskId) {
    const response = await this.auth.rest(
      "/reload-tasks/{taskId}".replace("{taskId}", taskId),
      {
        method: "GET"
      }
    );
    const responseJson = await response.json();
    return new Task(responseJson, this.auth);
  }
  /**
    Find and return the tasks that the user can access.
  
      @param appId - The case sensitive string used to search for a task by app ID.
      @param limit - The maximum number of resources to return for a request. The limit must be an integer between 1 and 100 (inclusive).
      @param next - The cursor to the next page of resources. Provide either the next or prev cursor, but not both.
      @param partial - The boolean value used to search for a task is partial or not
      @param prev - The cursor to the previous page of resources. Provide either the next or prev cursor, but not both.
    */
  async getReloadTasks(queryParams = { limit: 10 }) {
    const response = await this.auth.rest(
      `/reload-tasks?${new URLSearchParams(Object.entries(queryParams).filter(([, v]) => v !== void 0).map(([k, v]) => [k, v.toString()])).toString()}`,
      {
        method: "GET"
      }
    );
    const responseJson = await response.json();
    return new ListableResource(responseJson, this.auth, Task, queryParams);
  }
  /**
    Create a task for a specified app.
  
      @param data - Request body specifying the task parameters.
    */
  async create(data) {
    const response = await this.auth.rest(
      "/reload-tasks",
      {
        method: "POST",
        body: JSON.stringify(data)
      }
    );
    const responseJson = await response.json();
    return new Task(responseJson, this.auth);
  }
}
export {
  ReloadTasks,
  Task
};
